i6_models/primitives/specaugment.py

```diff
@@ -0,0 +1,118 @@
+    """
+    Mask tensor along axis using N in [min_num, max_num] masks of length [0, max_len]
+
+    :param tensor: e.g. [B, ..., A, ...] but arbitrary axis order
+    :param batch_axis: index of the batch axis
+    :param axis: which axis to mask
+    :param min_num: minimum number of masks
+    :param max_num: maximum number of masks
+    :param max_amount: mask length drawn uniformly from [0, max_amount]
+    """
+
+    batch_dim = tensor.shape[batch_axis]
+    num_masks = torch.randint(min_num, max_num, size=(batch_dim,)).to(device=tensor.device)  # [B]
+
+    z = -torch.log(-torch.log(torch.rand((batch_dim, tensor.shape[axis])).to(device=tensor.device)))  # [B,dim]
+    _, indices = torch.topk(z, num_masks.max().item(), dim=1)
+
+    for i in range(num_masks.max().item()):
+        tensor = _mask(tensor, batch_axis, axis, indices[:, i], max_len)
+    return tensor
+
+
+def zero_specaugment(
+    audio_features: torch.Tensor,
+    time_min_num_masks: int,
+    time_max_num_masks: int,
+    time_mask_max_size: int,
+    freq_min_num_masks: int,
+    freq_max_num_masks: int,
+    freq_mask_max_size: int,
+):
+    """
+    Specaugment from legacy rossenbach/zeineldeen/zeyer attention setups (usually called specaugment_v2.py or so),
+    but without any step-based scheduling. Fills masks with zeros.
+
+    Basically just a convenience wrapper around _random_mask.
+
+    :param audio_features: e.g. log-mel features as [B, T, F]
+    :param time_min_num_masks: minimum number of masks along T
+    :param time_max_num_masks: maximum number of masks along T
+    :param time_mask_max_size: maximum size of masks along T
+    :param freq_min_num_masks: minimum number of masks along F
+    :param freq_max_num_masks: maximum number of masks along F
+    :param freq_mask_max_size: maximum size of masks along F
+    :return: masked audio features
+    """
+    assert len(tensor.shape) == 3
+    tensor = _random_mask(
+        audio_features, 0, 1, time_min_num_masks, time_max_num_masks, time_mask_max_size
+    )  # time masking
+    tensor = _random_mask(
+        audio_features, 0, 2, freq_min_num_masks, freq_max_num_masks, freq_mask_max_size
+    )  # freq masking
+    return tensor
+
+
+def zero_specaugment_by_length(
+    audio_features: torch.Tensor,
+    time_mask_per_n_frames: int,
+    time_min_num_masks: int,
+    time_mask_max_size: int,
+    freq_min_num_masks: int,
+    freq_max_num_masks: int,
+    freq_mask_max_size: int,
+):
+    """
+    Convenience wrapper around zero_specaugment with time-length adaptive number of masks
+
+    :param audio_features: e.g. log-mel features as [B, T, F]
+    :param time_mask_per_n_frames: maximum number of masks depending on length T.
+        They are still drawn depending on the full batch length, so shorter sequences
+        might get more masks than that by chance, or none at all when all masks
+        fall into the padding space.
```

Comment:
The naming and documentation is not clear to me.

I would go with `mask_per_frame_ratio` or `frame_per_mask_ratio` (I prefer the first one, but then the implementation should be inverted and that might conflict with the expectations of people that have historically used this code...)

and then the description should be something  along "Insert (on average / at maximum) 1 mask for every frame_per_mask_ratio frames in the audio"

Reply:
I like frame per mask more because then you stay in the integer range... This is why I avoided ratio because for me this implies float. But of course we can change it to work like that. The original idea was just to copy the old code to PyTorch, but I am also not against reformulating it. But then there are additional changes I would like to add, like a flag for mean instead of zero replacement.

Reply:
By all means keep the integer and behavior as is i.e. frame_per_mask . I also do not insist on the word "ratio" in the variable name. 
I just found it confusing as by the description "maximum number of masks depending on .." I would have put in the maximum number of masks. But actually we want the number of frames per which a new mask might be inserted..

Reply:
Ah I understand, I will try to come up with something better

Reply:
> I will try to come up with something better

I have to say I do not really like the alternatives...